unidist/core/backends/mpi/core/monitor/loop.py

```diff
@@ -193,28 +210,72 @@ def monitor_loop():
+                )
+            reservation_info = shm_manager.get(request["id"])
+            if reservation_info is None:
+                reservation_info = shm_manager.put(request["id"], request["size"])
+                reservation_info["is_first_request"] = True
+            else:
+                reservation_info["is_first_request"] = False
+
+            communication.mpi_send_object(
+                mpi_state.comm, data=reservation_info, dest_rank=source_rank
+            )
+        elif operation_type == common.Operation.REQUEST_SHARED_DATA:
+            info_package = communication.mpi_recv_object(mpi_state.comm, source_rank)
+            data_id = info_package.pop("id", None)
+            if data_id is None:
+                raise ValueError("Requested DataId is None")
+            reservation_info = shm_manager.get(data_id)
+            if reservation_info is None:
+                raise RuntimeError(f"The monitor do not know the data id {data_id}")
+            sh_buf = shared_store.get_shared_buffer(
+                reservation_info["first_index"], reservation_info["last_index"]
+            )
+            communication.mpi_send_byte_buffer(
+                mpi_state.comm,
+                sh_buf,
+                dest_rank=source_rank,
+            )
+        elif operation_type == common.Operation.CLEANUP:
+            cleanup_list = communication.recv_serialized_data(
+                mpi_state.comm, source_rank
+            )
+            shm_manager.clear(cleanup_list)
         elif operation_type == common.Operation.READY_TO_SHUTDOWN:
             workers_ready_to_shutdown.append(source_rank)
-            shutdown_workers = (
-                len(workers_ready_to_shutdown) == mpi_state.world_size - 2
+            shutdown_workers = len(workers_ready_to_shutdown) == len(
+                mpi_state.workers
             )  # "-2" to exclude ``Root`` and ``Monitor`` ranks
+        elif operation_type == common.Operation.SHUTDOWN:
+            SharedStore.get_instance().finalize()
+            if not MPI.Is_finalized():
+                MPI.Finalize()
+            break  # leave event loop and shutdown monitoring
         else:
             raise ValueError(f"Unsupported operation: {operation_type}")
 
         if shutdown_workers:
-            for rank_id in range(
-                communication.MPIRank.FIRST_WORKER, mpi_state.world_size
-            ):
-                communication.mpi_send_operation(
-                    mpi_state.comm,
-                    common.Operation.SHUTDOWN,
-                    rank_id,
-                )
+            for rank_id in range(communication.MPIRank.ROOT + 1, mpi_state.world_size):
+                if rank_id != mpi_state.rank:
+                    communication.mpi_send_operation(
+                        mpi_state.comm,
+                        common.Operation.SHUTDOWN,
+                        rank_id,
+                    )
             communication.mpi_send_object(
```

Comment:
Will all monitors be sending a request?

Reply:
only root monitor will send this requst, because the "shutdown_workers" condition is true only for root monitor. Other monitors will receive shutdown operation and will finalized